### ISSUE TITLE
Security Advisory Features

### ISSUE THREAD

**Issue by naderman** [composer/composer#10329]
For a while now we have had security advisories available through an API on packagist.org: https://packagist.org/apidoc#list-security-advisories. Packagist.org also lists advisories on packages and marks vulnerable versions with a warning sign. However this information is not easily accessible in Composer itself. A common solution to avoid installing vulnerable versions is requiring the [roave/security-advisories](https://github.com/Roave/SecurityAdvisories) package which conflicts with all the packages in https://github.com/FriendsOfPHP/security-advisories. The same database is imported into packagist.org (with additional ones coming, e.g. https://github.com/composer/packagist/pull/1206). There are also various tools available which scan repositories either on the CLI or as a service using the same database and/or others (e.g. https://github.com/symfonycorp/security-checker-action).

As updating vulnerable versions and identifying them is a core task of working with dependencies, Composer should have tools built in to deal with them. In particular:

- `composer audit` command to display a list of vulnerable versions with details on the advisories used in the current composer.lock. Should ideally also tell the user what they need to update to get to safe versions, e.g. suggest a partial update command?
- `composer update` should allow filtering out any vulnerable versions. There are a two main alternatives
  - Keep the default behavior as-is, add a new option, e.g. `--filter-vulnerable`
  - Change the default behavior to filter out any vulnerable version, and add an option to install them anyway, e.g. `--allow-vulnerable`
  
  Handling partial updates will likely require additional options: Should locked versions remain or should they also get updated to a secure release? Should there be warnings printed about vulnerable versions in this case?
- Print warnings about vulnerable installed versions on all updates and installs.

The vulnerabilities should be readable from additional composer repositories as well, so that repositories like drupal's or magento's can publish their own advisories if they wish.

**Comment by naderman** [comment]
If we build this into composer update/install we may need to consider adding this metadata to the information packagist.org provides by default, instead of contacing an additional endpoint.

**Comment by dgfmdrv** [comment]
Thank you for working on this. 

For example yarn audit shows:
![image](https://user-images.githubusercontent.com/1451834/145985461-79058920-fd28-40e5-b095-b3e1973eea73.png)

**Comment by naderman** [comment]
@Seldaek There are actually additional features listed in this issue which are not yet implemented, so reopening.

**Comment by naderman** [comment]
I think it'd make a huge difference. As usual people will ignore warnings. Making Composer skip known-vulnerable versions by default is going to have a very positive impact. I don't disagree it's more work/complex, but I do think that's what we should really aim for.

**Comment by stof** [comment]
But this would probably require first to dump the advisory metadata as part of the composer repository instead of relying on a live API, because this would increase a lot the volume of requests done to packagist.org

**Comment by GuySartorelli** [comment]
> this would increase a lot the volume of requests done to packagist.org

I don't think it would need to - you could just get a list of advisories for all the packages being/currently installed in a single request near the start of the update/install command instead of at the end. Currently the default is to call the packagist API on every `composer update`, `composer install`, etc and since you can request advisories for all packages you're interested in at once, you can just move the timing of that API call to be as soon as you know what packages you're interested in.

**Comment by tedbow** [comment]
It seems that now `composer audit` will use the dynamic fetched documented in https://packagist.org/apidoc#list-security-advisories if `api-url` is set but otherwise it will use `security-advisories` section of an individual .json files for packages.

I am working on [PHP-TUF Composer integration plugin](url)(Implementing [The Update Framework](url)) and we are trying to figure how to handle this. drupal.org will be our first use case(unless somebody else beats us to it 😉).

drupal.org has provided this endpoint https://www.drupal.org/project/project_composer/issues/3399877 see example https://packages.drupal.org/8/security-advisories?packages[]=drupal/tfa&packages[]=drupal/swiftmailer

The problem from a TUF perspective is we protect all the other Composer metadata which reside in files but I think it would be impossible to protect the dynamic calls to `/security-advisories?packages[]=vendor/package` because this response needs to be dynamically created because each request could obviously have a unique set packages or other query parameters for each project.

Is there a suggested way to force `composer audit` to not use `/security-advisories?packages[]=vendor/package`  even if `api-url` is provided by the repository?

**Comment by Seldaek** [comment]
Does this really matter tho? Ensuring package integrity is great but signing this API seems kinda pointless.

And no, there is no way to prevent this right now, if an API is available it will use it because the way we have advisories stored in packagist.org for efficiency the static files do not contain the entire advisory information, so it is only fetched when someone actually runs an audit via the API, and otherwise only minimal info about which versions are affected are present in the metadata json files, e.g. from https://repo.packagist.org/p2/composer/composer.json
<img width="455" alt="image" src="https://github.com/composer/composer/assets/183678/54c5bba3-5a0f-4ede-9996-ecbc87d9f1d7">

Including the full details of advisories might be very long otherwise, especially for repos that have lots historical advisories.

**Comment by tedbow** [comment]
> Does this really matter tho? Ensuring package integrity is great but signing this API seems kinda pointless.

Well  I think it does create a big attack vector.

For instance TUF projects against [Freeze attacks](https://theupdateframework.io/security/) where 

> An attacker continues to present files to a software update system files that the client has already seen. As a result, the client is kept unaware of new files.

Because all of our Composer metadata is protected by TUF we should protected from this. We should know if the versions of particular package is not the most up-to-date because there is man-in-middle attack for instance.

But since which package versions are **insecure** will not be protected by TUF(because of the dynamic nature)  therefore that list is open to attacks that TUF protects against. Therefore the system would be more vulnerable to an attack where it is fooled into thinking the current version is **not** insecure when it actually is or into allowing an update where the update to intended package actually brought in version of dependency that is fact insecure itself.

I am not saying this is deal breaker for our use of TUF, regardless of this problem TUF when used in the Drupal context for now will not be a fully protected by TUF, because of our need for dependencies that aren't TUF protect and our need for metadata about projects that is currently only available in non-TUF sources
